nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,418 @@
+          description = "A tool to seamlessly migrate previous configuration of Netbird";
+
+          before = [ "${client.name}.service" ];
+          requiredBy = [ "${client.name}.service" ];
+
+          serviceConfig = {
+            ExecStart = getExe (pkgs.writeShellApplication {
+              name = "${client.name}-migrate";
+              runtimeInputs = with pkgs; [ coreutils findutils jq ];
+              runtimeEnv = client.environment // {
+                hardened = toString client.hardened;
+                service_name = client.name;
               };
+              text = ''
+                dirs=(
+                  # see https://github.com/netbirdio/netbird/issues/1079#issuecomment-1823180247
+                  "/var/cache/$service_name"
+                  "/etc/$service_name"
+                  "/var/log/$service_name"
+                  "/var/run/$service_name"
+                  "/var/lib/$service_name"
+                )
 
-              stopIfChanged = false;
-            }
-          )
-          cfg.tunnels;
-    })
+                if [[ "$hardened" == 1 ]] ; then
+                  # find the first folder not owned by root
+                  for dir in "''${dirs[@]}"; do
+                    [[ -e "$dir" ]] || continue
+                    owner="$(stat -c "%u:%g" "$(realpath "$dir")")"
+                    [[ "$owner" == "0:0" ]] || break
+                  done
+                else
+                  owner=0:0
+                fi
+
+                if [[ -z "$owner" ]] ; then
+                  echo "OK: no ownership information found, probably first run" >&2
+                  exit
+                fi
+
+                echo "Checking hardened=$hardened specific permissions..." >&2
+
+                ensure_owner() {
+                  local dir="$1" repr="$2"
+                  echo "Checking ownership of $repr" >&2
+
+                  while IFS= read -r file; do
+                    if [ "$(stat -c "%u:%g" "$file")" == "$owner" ] ; then
+                      echo "- OK:     $file already owned by $owner" >&2
+                    else
+                      echo "- CHANGE: $file changing owner to $owner" >&2
+                      chown "$owner" "$file"
+                    fi
+                  done < <(find "$dir")
+
+                }
+
+                for dir in "''${dirs[@]}" ; do
+                  [[ -e "$dir" ]] || continue
+
+                  repr="$dir"
+                  realdir="$(realpath "$dir" || echo -n "$dir")"
+
+                  if [ "$dir" != "$realdir" ] ; then
+                    ensure_owner "$dir" "$repr"
+                    repr="$dir (-> $realdir)"
+                    dir="$realdir"
+                  fi
+                  ensure_owner "$dir" "$repr"
+                done
+              '';
+            });
+            Type = "oneshot";
+            User = "root";
+            RemainAfterExit = "yes";
+            Restart = "no";
+          };
+        }
+      );
```

Comment:
I don't really get what you are trying to do with this whole bash blob, systemd should set the permissions correctly at startup

Reply:
it does so only for the first start of the service, not when toggling between hardened/not hardened

Reply:
actually seems like it does the right thing as long as the `Directory` itself is owned by the "wrong" user

Reply:
So as for now everything is owned by root, systemd will do it for us automatically

Reply:
yeah, I'm confirming it on my branch, but got some stuff messed up due too many open PRs (merged into that branch)

Reply:
It seems to work for all directories except `ConfigurationDirectory`